lib/network/default.nix

```
@@ -0,0 +1,107 @@
+{ lib }:
+{
+  ipv6 = {
+    /**
+      Normalize an IPv6 address by removing the "::" compressions and padding
+      them with the necessary number of zeros to create the correct address. Converts
+      an address from the string to the list of strings which then can be parsed using
+      `_parseNormalized`. Throws an error when the address is malformed.
+
+      # Type: String -> [ String ]
+    */
+    _normalizeIpv6 =
+      addr:
+      if builtins.match "^[0-9A-Fa-f:]+$" addr == null then
+        throw "${addr} contains malformed characters for IPv6 address"
+      else
+        let
+          doubleColon = "::";
+          piecesMax = 8;
+          pieces = lib.strings.splitString ":" addr;
+          piecesNoEmpty = lib.lists.remove "" pieces;
+          piecesNoEmptyLen = builtins.length piecesNoEmpty;
+          emptyCnt = builtins.length pieces - builtins.length piecesNoEmpty;
+          hasPrefix = lib.strings.hasPrefix doubleColon addr;
+          hasSuffix = lib.strings.hasSuffix doubleColon addr;
+          hasInfix = lib.strings.hasInfix doubleColon addr;
+          emptyExpected =
+            if hasPrefix || hasSuffix then
+              2
+            else if hasInfix then
+              1
+            else
+              0;
+        in
+        if addr == doubleColon then
+          builtins.genList (_: "0") piecesMax
+        else if
+          emptyCnt != emptyExpected
+          || (hasInfix && piecesNoEmptyLen >= piecesMax)
+          || (!hasInfix && piecesNoEmptyLen != piecesMax)
+        then
+          throw "${addr} is incorrect IPv6 address"
+        else if hasPrefix || hasSuffix then
+          let
+            zeros = builtins.genList (_: "0") (piecesMax - piecesNoEmptyLen);
+          in
+          if hasPrefix then zeros ++ piecesNoEmpty else piecesNoEmpty ++ zeros
+        else if hasInfix then
+          builtins.concatMap (
+            el: if el == "" then builtins.genList (_: "0") (piecesMax - piecesNoEmptyLen) else [ el ]
+          ) pieces
+        else
+          pieces;
+
+    /**
+      Converts hexadecimal strings to base-10 numbers. Returns an error if the
+      number cannot be parsed or if the number is not a 16-bit unsigned integer.
+
+      # Type: [ String ] -> [ Int ]
+    */
+    _parseNormalized =
+      addr:
+      let
+        u16FromHexStr =
+          hex:
+          let
+            parsed = lib.fromHexString hex;
+          in
+          if
+            0 <= parsed && parsed <= 65535 # 2^16 - 1
+          then
+            parsed
+          else
+            throw "0x${hex} is not valid u16 integer";
+      in
+      assert lib.asserts.assertMsg (
+        builtins.length addr == 8
+      ) "_parseNormalized expected to recieve normalized IPv6 address";
+      builtins.map (el: u16FromHexStr el) addr;
```

```suggestion
      builtins.map (piece: u16FromHexStr piece) addr;
```

We should keep the same language everywhere to avoid confusion.